pkg/activator/revision_backends.go

```diff
@@ -0,0 +1,285 @@
+package activator
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strconv"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/pkg/controller"
+	"knative.dev/serving/pkg/apis/networking"
+	"knative.dev/serving/pkg/apis/serving"
+	"knative.dev/serving/pkg/network"
+	"knative.dev/serving/pkg/network/prober"
+	"knative.dev/serving/pkg/queue"
+	"knative.dev/serving/pkg/reconciler"
+)
+
+// RevisionDestsUpdate is the main output of a RevisionBackendsManager which contains the l4 dests
+// information for how to reach a revision via PodIP
+type RevisionDestsUpdate struct {
+	Rev   RevisionID
+	Dests []string
+}
+
+const (
+	probePortName  string        = "http"
+	probeTimeout   time.Duration = 300 * time.Millisecond
+	probeFrequency time.Duration = 200 * time.Millisecond
+)
+
+type revisionWatcher struct {
+	rev          RevisionID
+	updateCh     chan<- *RevisionDestsUpdate
+	dests        []string
+	healthStates map[string]bool
+
+	transport http.RoundTripper
+	destsChan <-chan []string
+	logger    *zap.SugaredLogger
+}
+
+func newRevisionWatcher(rev RevisionID, updateCh chan<- *RevisionDestsUpdate,
+	destsChan <-chan []string, transport http.RoundTripper,
+	logger *zap.SugaredLogger) *revisionWatcher {
+	return &revisionWatcher{
+		rev:          rev,
+		updateCh:     updateCh,
+		healthStates: make(map[string]bool),
+		transport:    transport,
+		destsChan:    destsChan,
+		logger:       logger,
+	}
+}
+
+func endpointsToDests(endpoints *corev1.Endpoints) []string {
+	ret := []string{}
+
+	for _, es := range endpoints.Subsets {
+		portVal := int32(0)
+		for _, port := range es.Ports {
+			if port.Name == probePortName {
+				portVal = port.Port
+				break
+			}
+		}
+
+		// This endpoint doesnt have a port we want to probe
+		if portVal == 0 {
+			continue
+		}
+
+		for _, addr := range es.Addresses {
+			// Prefer IP as we can avoid a DNS lookup this way
+			ret = append(ret, net.JoinHostPort(addr.IP, strconv.Itoa(int(portVal))))
+		}
+	}
+
+	return ret
+}
+
+func filterHealthyDests(dests map[string]bool) []string {
+	ret := make([]string, len(dests))
+	i := 0
+	for dest, healthy := range dests {
+		if healthy {
+			ret[i] = dest
+			i++
+		}
+	}
+	return ret[:i]
+}
+
+type destHealth struct {
+	dest   string
+	health bool
+}
+
+func (rw *revisionWatcher) checkDests() {
```

Comment:
I think the assumption is that this is not called concurrently? Maybe add that in a comment.